### ISSUE TITLE
Composer update is deleting pathed Windows symlinked local repos again, ref #4955

### ISSUE THREAD

**Issue by narration-sd** [composer/composer#7025]
Looks exactly like #4955 --- do  you  have a regression?  

Very disconcerting, only saved because just had pushed the github repo to an origin.

My `composer.json`:

{
  "name": "craftcms/craft",
  "description": "Craft CMS",
  "keywords": [
    "craft",
    "cms",
    "craftcms",
    "project"
  ],
  "license": "MIT",
  "homepage": "https://craftcms.com/",
  "type": "project",
  "support": {
    "email": "support@craftcms.com",
    "issues": "https://github.com/craftcms/cms/issues",
    "forum": "https://craftcms.stackexchange.com/",
    "source": "https://github.com/craftcms/cms",
    "docs": "https://craftcms.com/docs",
    "rss": "https://craftcms.com/changelog.rss"
  },
  "minimum-stability": "beta",
  "repositories": [
    {
      "type": "path",
      "url": "plugin-dev/live-vue",
      "symlink": true
    }
  ],
  "require": {
    "php": ">=7.0.0",
    "craftcms/cms": "dev-develop as 3.0.0-RC99",
    "vlucas/phpdotenv": "^2.4.0",
    "roave/security-advisories": "dev-master",
    "craftcms/element-api": "^2.5",
    "craftcms/redactor": "^1.0.0-beta.1",
    "laravel/homestead": "^7.0.0",
    "narrationsd/live-vue": "^0.8.0"
  },
  "autoload": {
    "psr-4": {
      "modules\\": "modules/"
    }
  },
  "config": {
    "optimize-autoloader": true
  },
  "scripts": {
    "post-root-package-install": [
      "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
    ],
    "post-create-project-cmd": [
      "@composer dump-autoload -o",
      "@php craft setup/welcome"
    ]
  },
  "require-dev": {
  }
}

Output of `composer diagnose`:

C:\vagrant\gnu\seventh-page_html>composer diagnose
Checking composer.json: WARNING
require.roave/security-advisories : unbound version constraints (dev-master) should be avoided
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys:
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.6.2
PHP version: 7.1.13
PHP binary path: C:\php\php.exe

When I run this command:

composer update (with narrationsd/live-vue at version 0.8.1)

I get the following output:

C:\vagrant\gnu\seventh-page_html>composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating narrationsd/live-vue (0.2.1 => 0.8.1):     Update failed (RecursiveDirectoryIterator::__construct(C:\vagrant\gnu\seventh-page_html\
vendor/narrationsd/live-vue,C:\vagrant\gnu\seventh-page_html\vendor/narrationsd/live-vue): The system cannot find the path specified. (code: 3))

    Would you like to try reinstalling the package instead [yes]?

***** At this point, not answering the prompt,  the entire folder pathed for  narrationsd/live-vue utterly and silently disappeared! *****

And I expected this to happen:

Of course it should have simply updated to the 0.8.1 version in all composer and autoload regards.

**Comment by curry684** [comment]
I just had a quick look and all the code I wrote for the junctions in Feb 2016 is still there verbatim in master and 1.6.2 without any changes since then. I'll admit right away that's no guarantee of anything as the detection of junctions during uninstall is pretty hacky (it depends on bitmasks of low level filesystem flags), and I would not at all be surprised if some Windows or PHP update would subtly change the mechanism along the way.

Regrettably I'm no longer developing on Windows so it's pretty hard for me to dive into the issue. Are you in a position to debug this yourself? I'll gladly give the required pointers if you are up to it.

**Comment by curry684** [comment]
You'd have to start with a breakpoint here: https://github.com/composer/composer/blob/master/src/Composer/Util/Filesystem.php#L653

The big comment underneath that line illustrates my point about the mechanism having some volatile aspects   ;)

**Comment by curry684** [comment]
On a separate note: I'd be in favor of rewriting all the junction code altogether. I wrote the subsystem early 2016 because it was a much requested usecase for Windows-based developers, but the inherent problem was that NTFS Symlinks were not practically usable because they required administrative rights, and Composer is just about the last tool you want to use as any kind of root all the time (we don't show the big fat warnings if you do without good reason). But the [Windows 10 Creators Update has lifted this limitation](https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10/#Ed9Olhkz6hJp4KWV.97), so by now we could switch to real symlinks instead.

Caveat would likely be that I'd still have to be highly specific code as I highly doubt PHP's native `is_symlink` call and its siblings properly recognize Windows symlinks, especially in EOL PHP versions that are still supported by Composer.

**Comment by narration-sd** [comment]
Ah, my, Niels....and  thanks for your copious info around thisd

I'm using the latest php 7.1, along with the latest Win10 having the new threat patches in it, so it's very possible one of these just disturbed shakey foundations under junction-as-symlink handling.

I'm afraid I'm not up to being a troubleshooter on this here. I do feel it's likely to bite severely, though.

I wonder if it would be a good idea for Composer to just refuse on Windows composer updates, if pathed, symlinked repos are present?

A workaround I used was to delete the entire /vendor, then do a composer install. So another possibility would be for Composer, in the case of updates w/Win 'symlinks', to just do the equivalent, but only for the locally pathed repo. 

Here's how it might do that.  I saw it left the junction/symlink present, after destroying what  it pointed to. So, what if it deleted the symlink first, under the bad situation known, then simply re-installed the pathed local repo, as a substitute for 'updating' it? Keeping up its book-keeping as far as the usual composer-only portions of /vendor, and of course composer.lock?

It seems to me this could be a viable solution, would be hardly noticed in a performance  sense, and would get you yourself as well as Composer out from under a very tricky problem entirely here, no?

As farr as winbash etc. or whatever it's called these days, I don't trust the symlink handling in the slightest, at least from what I remember reading about it. I have enough trouble with it just using it to run ansible and a small modicum of docker tools, which are always severely out of date there, to build and install some not too complicated containers (not involved in project here).

What do you think about that line of solution?  And please don't stay up too late about this, very conscious of Euro timezone as used to live there....

**Comment by narration-sd** [comment]
p.s. if there isn't something else the matter with my proposal, it seems it would also clean up some mis-behaviour I seem to have a foggy memory about, in terms of missing some info on updates from local repos. 

Sorry I can't seem to be more specific about that, but workaround was to remove and re-require the local repo. Which is sort of similar also to what I propose above.

**Comment by curry684** [comment]
> So another possibility would be for Composer, in the case of updates w/Win 'symlinks', to just do the equivalent, but only for the locally pathed repo.

That's a pretty good description of how the mechanism works  😉 

I linked to the `Filesystem::isJunction` function specifically. The problem is that recursively deleting a directory requires recursively deleting all its contents first. The default implementations rely on `is_symlink` not to do exactly what you are experiencing, which does not detect junctions at all -
technically those are low-level filesystem forks comparable to *nix hardlinks. So default Composer (and `Symfony/Filesystem` for all that matters) behavior is to recursively delete the entire library folder *including anything linked via junctions as it cannot detect them*. The `isJunction` function is there to detect this rather unwanted case and trigger `removeJunction` instead.

@Seldaek you still developing on Windows these days?

**Comment by narration-sd** [comment]
Oh my. Yes, I see. M. Seldaek to the rescue, we may hope....!

**Comment by narration-sd** [comment]
Hmm, @curry684 ...

Niels, I looked at your code and comment, then checked around a bit, ran a test.  

It _appears_ that a simple mod to your test would be more reliable than checking via some raw bit masks which might have changed.  This would be changing to simply comparing the mode values from PHP stat() and lstat().

You can see what you think -- below is the test of what those mode returns look like -- pretty clear. You'll have to plug in your own path for a junctioned Windows folder --  I used the vendor folder for my code that composer has been blowing away.

There do seem to be a number of not well explained or actualliy correct suggestions out there -- in particular the one about using array_diff won't work, as it only compares keys. But this one is simple, direct, and maybe it is correct? 

At least here, it looks like it would save the day, and using a vcs repo in Composer turned out very uncomfortable for develop/debugging -- big delays, and doesn't work correctly with PhpStorm ide debugging.

here's pathck.php -- and I learned the hard way about clearstatcache()...

 ```
<?php

function ckmodes ($path) {
    echo 'checking ' . $path . "\n";
    echo 'stat mode ' . var_export(stat($path)['mode'], true) . "\n";
    echo 'lstat mode ' . var_export(lstat($path)['mode'], true) . "\n\n";
}

function is_junction($path)
{
    clearstatcache(true, $path); // safety
    $stat_mode = stat($path)['mode'];
    clearstatcache(true, $path); // absolutely necessary
    $lstat_mode = lstat($path)['mode'];

    return strcmp($stat_mode, $lstat_mode) !== 0;
}

function test_junction($path) {
    echo $path . (is_junction($path) ? " is " : " is not ") . "a junction\n";
}

$junctiondir = 'vendor/narrationsd/live-vue';
$junctionedfile = 'vendor/narrationsd/live-vue/composer.json';

echo "\nfirst, folders: \n\n";
ckmodes($junctiondir);
ckmodes('.');  // local directory

echo "now, files:\n\n";
ckmodes($junctionedfile);
ckmodes($argv[0]);  // self

echo "finally, truths:\n\n";
test_junction($junctiondir);
test_junction('.');
test_junction($junctionedfile);
test_junction($argv[0]);
```

**Comment by curry684** [comment]
Seems plausible, make a PR I'd say ;)

And I'd still be in favor of switching to real symlinks though.

**Comment by Seldaek** [comment]
I am still on Windows yes, but I won't have time to investigate this for a bit I expect, so if you do have a solution that works @narration-sd please do open a PR.

**Comment by narration-sd** [comment]
Hmm Hmm Hmmm,

Guys, I wasn't feeling really fascinated by putting up a relatively untested 'fix', so more hours on this, which turned up 'fascinating' results.

-  first of all, at least the current 7.1 php lstat gives a 0 for ['mode'] for junctions (!)
- this would mean that the test at the end of isJunction() against & bits would always fail, giving you the right answer, for not quite correct reasons(?) on junctions.
- non-junction lstat modes do give bits, and provably not the junction ones, so again your isJunction() works.
- I've actually tested your copied full isJunction() routine in an expanded tester pasted in below; it works.
- this tester gives many details. Including that array_diff or array_diff_assoc does actually work as php doc proposes, just not as they say, as it gives two elements for the same single diff. But irrelevant to our case.
- I tried several other things
- using the phar instead of the Windows-installed composer equally destroys my win-junctioned module
- this is not surprising, since they are both reporting the same version, 1.6.2 and date
- I've tested (and printed the constant it looks for) your Platform::isWindows(), both in a clone of the composer git, and in the composer you bury inside of /vendor. Which even after updates is still verion 1.5.6 -- why??, Anyway, the constant is definitely there from php
- my tester gives the intermediate call results you make in isJunction(), which look right, and prove out once I added the copy of isWindows itself.
- I didn't have luck actually running my git clone bin/composer; it presumes to be installed where I won't put it, inside the /vendor/composer folder of my project where you already have the 1.5.6. This and time are putting an end to my experiments.

So, where are we?
- every time, and any version 1.6.2 install type, composer will zap what it junction-'soft linked' of my pathed code repos
- this doesn't at all appear now to be the fault of isJunction() failing, actually, as it appears to work, if for not quite the expected reasons.
- so, what is it???
- thus I haven't any PRs to propose, and would really feel out of court to do this, having not any of the range of testing ability, phar ability, etc, etc.. 
- I might have if it had turned out to be a simple one-liner to improve isJunction(), but that's not what's needed, and would in any case be at all comfortably for a huge community of use.
- I really sympathize with the idea to make this whole area operate with the later NTFS true-?-er?-symlinks, but it sounds very fraught, re: Microsoft, and re: community who may still be using pre-Vista-is-it PCs. Lots of corp and military still on XP. And would that be Microsoft-symlinks-as-php-versions-may-or-not-properly-reflect-it? Or Windows code, but then what about composer.phar? Yikes, to use an Americanism....and anyway, not available today.

So, honestly, i think fixing the reported code-zap issue needs the people who are really set up with the source and testing and understanding to track it down. It's a very hard fault on Win10 latest everything and php 7.1, Windows composer.exe, php composer.phar, and presumably composer-from-git, given a way to actually run that past the point where it can't find its expected /vendor files.

I will get around it for the present by  care to have my source all committed to remote repo, every time  before doing composer update, if the problem only occurs when I update the version in my code's composer.json. And also just removing the junction you made for that code in /vendor/myarea before composer update, as that safeties, and composer update will remake it. 

I don't have to tell you these procedural needs put a kind of hotness on this issue.

Extra points: why is that seemingly unused (it doesn't print var_dumps I put in it on normal composer operations, just when I call it myself directly).  And why does Composer update it as I've seen to an earlier than latest version; something in Packagist?

That's  my Sunday afternoon, and here's that more-copious tester - pathck.php

```
<?php

function ckmodes ($path) {
    echo 'checking ' . $path . "\n";
    clearstatcache(true, $path); // safety
    $stat_full = stat($path);
    clearstatcache(true, $path); // absolutely necessary
    $lstat_full = lstat($path);


    echo 'stat mode ' . var_export(stat($path)['mode'], true) . "\n";
    echo 'lstat mode ' . var_export(lstat($path)['mode'], true) . "\n\n";
    echo 'array_diff ' . var_export(array_diff_assoc($stat_full, $lstat_full), true) . "\n\n";;
    echo 'is_dir ' . (is_dir($path) ? 'yes' : 'no')
        . ', is_link ' . (is_link($path) ? 'yes' : 'no')
        . ', mode & 0xC000 ' . ($lstat_full['mode'] & 0xC000) . "\n\n";
}

function stat_test_junction($path)
{
    clearstatcache(true, $path); // safety
    $stat_mode = stat($path)['mode'];
    clearstatcache(true, $path); // absolutely necessary
    $lstat_mode = lstat($path)['mode'];

    return strcmp($stat_mode, $lstat_mode) !== 0;
}

function zero_lstat_test_junction ($path) {
    clearstatcache(true, $path); // absolutely necessary
    $lstat_mode = lstat($path)['mode'];

    return !($lstat_mode & 0xC000) /*|| ($lstat_mode === 0)*/;
}

function test_junction($path) {
    echo $path . (zero_lstat_test_junction($path) ? " is " : " is not ") . "a junction\n";
}

function isWindows()
{
//    \var_dump('here in vendor: ' . PHP_WINDOWS_VERSION_BUILD
//        . (defined('PHP_WINDOWS_VERSION_BUILD') ? ' yes' : ' no'));
    return defined('PHP_WINDOWS_VERSION_BUILD');
}

function isJunction($junction)
{
    if (! /* Platform:: */ isWindows()) {
        return false;
    }
    if (!is_dir($junction) || is_link($junction)) {
        return false;
    }
    /**
     * According to MSDN at https://msdn.microsoft.com/en-us/library/14h5k7ff.aspx we can detect a junction now
     * using the 'mode' value from stat: "The _S_IFDIR bit is set if path specifies a directory; the _S_IFREG bit
     * is set if path specifies an ordinary file or a device." We have just tested for a directory above, so if
     * we have a directory that isn't one according to lstat(...) we must have a junction.
     *
     * #define	_S_IFDIR	0x4000
     * #define	_S_IFREG	0x8000
     *
     * Stat cache should be cleared before to avoid accidentally reading wrong information from previous installs.
     */
    clearstatcache(true, $junction);
    $stat = lstat($junction);

    return !($stat['mode'] & 0xC000);
}

function orig_test_junction($path) {
    echo $path . (isJunction($path) ? " is " : " is not ") . "a junction\n";
}


$junctiondir = 'vendor/narrationsd/live-vue';
$junctionedfile = 'vendor/narrationsd/live-vue/composer.json';

echo "\nfirst, folders: \n\n";
ckmodes($junctiondir);
ckmodes('.');  // local directory

echo "now, files:\n\n";
ckmodes($junctionedfile);
ckmodes($argv[0]);  // self

echo "finally, truths:\n\n";
test_junction($junctiondir);
test_junction('.');
test_junction($junctionedfile);
test_junction($argv[0]);

echo "\nbut which ones, truths; let's see the original iJunction's result:\n\n";
orig_test_junction($junctiondir);
orig_test_junction('.');
orig_test_junction($junctionedfile);
orig_test_junction($argv[0]);
```

**Comment by narration-sd** [comment]
just a wild thought after. What if current Win10, especially after some things highly touted that they;ve recently done for the winbash/WSL symlinks, is actually making proper NTFS symlinks where you think you're making junctions???

Or otherwise lousing things up. I mentioned I wasn't happy with this WSL update anyway, as what they did is still not compatible with rsync/ansible file transfers; not the same as unix, so you are driven to copies.

This raises the question of whether the fact that I've got WSL installed has anything to do with the composer pathed repo-zapping. I'll consider deinstalling it after I have a look at how much to rebuild if I did it, as deinstall destroys everything your actually-in-WSL files.

And even if WSL is the culprit, I would think changes they made to support it would be in the native NTFS filesystem, not something patched on. Of course, Microsoft...

But if this is so, WSL-install-provoked, that would substantially reduce your attack surface for this Composer issue.

**Comment by narration-sd** [comment]
Ok, a rapid reply on the last thought.

- I managed to copy out everything vital from WSL -- except what I now realize I forgot, but I have other copies for those.
- I took Microsoftt WSL off the machine, with a reboot; verified it was gone. After the reboot, any WSL-installed NTFS patches would be gone.
- I tested and again blew away my module repo by simply upping it's composer.json version, then composer update on the master project. Repaired via git origin.

So, 
- this issue's problem's are not in patches actually having WSL/winbash being installed
- but it could still be due to NTFS changes due to Microsoft's highly touted 'now symlinks on WSL fully work' having made mods on the Win10 update that accompanied this, not many months ago by memory, though it could have been a spring update, not sure.
- or, of course later php7.1 etc. update levels treatment of NTFS, possibly accompanying any of this
- at least I think we can understand isJunction() works. But the rest of the armature around your making of junctions, and whether that's actually now what they really are, and not NTFS symlinks, giving wrong answers?  Because these soft links were made by composer...

Enough, and late dinner now. Cheers to both of you.
C.

**Comment by curry684** [comment]
Thanks for the help and the elaborate research :)

> What if current Win10, especially after some things highly touted that they've recently done for the winbash/WSL symlinks, is actually making proper NTFS symlinks where you think you're making junctions???

It should still work fine as it always checks `is_symlink` first still.

I'll try to hammer up a PR to disable functioning asap in preparation of a real fix.

**Comment by Seldaek** [comment]
@narration-sd what version of windows do you have exactly? Because I am on `10.0.16299 Build 16299` with WSL installed and I do not experience this problem as far as I can tell with junctioned path repositories.

**Comment by narration-sd** [comment]
Wow. It seems the same to this level; Win 10 64 Home v1709 16299.192, with the patches for the Meltdown/Spectre vuln the last am aware of installed. 

Php is 7.1.13 from 5 Jan,, nts-x64 -- had to back off from 7.2 for some small reasons. I don't think I have anything unusual enabled as far as extensions, only a few like curl.

There is that small issue mentioned, that the Win exe of Composer I usually use, and the phar I tried today are both version 1.6.2, while the one hidden in the /vendor folder is for some reason 1.5.6 even though I believe to have seen it updating in last few days. Not sure what that is ever used for (and curios).

Scratching my head; machine itself is very stable, 16Gb, Samsung SSD, i7 which seldom breathes hard unless something runs away....

**Comment by narration-sd** [comment]
Also on a hunch I just tried running the same version-changed module, composer update sequence, but from a straight Windows cmd shell instead of the bash I usually use, and the same deletion as ever happened.

So I tried the same thing with another project pathing a more elaborate plugin(module) repo. Exactly the same failure, blowing away that somekind-linked folder of code entirely, recovered as usual via git origin pull.

In both these cases now I did a composer update without any local version changes, which resulted in a big pull of the underlying cms, where I'm on the origin develop repo over Packagist. No problems, even the one which does a stash-reapply over my outstanding pr there. But this local pathed repo on the next composer update, boom.  

Here's what it looks like, not that I think there's any extra information..

```
C:\vagrant\gnu\dpprime_html>composer update
Loading composer repositories with package information
Reading composer.json of selvinortiz/doxter (feature/toolbar-plugin-field-settinReading composer.json of selvinortiz/doxter (master)                            Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating narrationsd/standardsagent (2.0.4 => 2.0.5):     Update failed (RecursiveDirectoryIterator::__construct(C:\vagrant\gnu\dpprime_html\vendor/narrationsd/standardsagent,C:\vagrant\gnu\dpprime_html\vendor/narrationsd/standardsagent): The system cannot find the path specified. (code: 3))
    Would you like to try reinstalling the package instead [yes]? n
The following exception may be caused by a stale entry in your cmd.exe AutoRun
Check https://getcomposer.org/doc/articles/troubleshooting.md#-the-system-cannot-find-the-path-specified-windows- for details


  [UnexpectedValueException]
  RecursiveDirectoryIterator::__construct(C:\vagrant\gnu\dpprime_html\vendor/
  narrationsd/standardsagent,C:\vagrant\gnu\dpprime_html\vendor/narrationsd/s
  tandardsagent): The system cannot find the path specified. (code: 3)


update [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--lock] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [--with-dependencies] [--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-i|--interactive] [--root-reqs] [--] [<packages>]...


C:\vagrant\gnu\dpprime_html>
```

**Comment by curry684** [comment]
I've set up a fresh Windows dev env on completely up to date Win10 with PHP 7.2.1, no issue at all, working as intended, with explicit output:
```
  - Removing junction for omines/datatables-bundle (dev-master)
```
If you can set up or concisely describe an isolated [SSCCE](http://sscce.org/) I'll be grateful, I need a reproduction to be able to do anything.

**Comment by narration-sd** [comment]
Yes, of course I understand, Niels. 

I'll give it a shot, not tonight as it is already 0200.

**Comment by curry684** [comment]
Well no rush anymore if it isn't deleting code for everyone all over the world anymore   ;)

**Comment by narration-sd** [comment]
This is true ;)

My mind is kind of churning in visualization over what could possibly be unique here - and changed over the many months I've been successfully working these project repos with composer. 

The only thing I can come up with as possibly unusual is that the pathed modules are actually git submodules into the main repo level composer is operating from. 

That, and the extra composer full source within /vagrant/composer/composer, possibly an artifact of the complex Craft CMS self-update system, which might explain how it gets to be a slightly of date, and coming from Packagist. It's not in play in this composer update, would be pretty sure.

Well, one more thing. I'll check it and be back if it is affecting matters -- PhpStorm with its constant file-watching has been running on the project of interest every time...

**Comment by narration-sd** [comment]
...and nope, bang just the same with no extra parties running....

**Comment by narration-sd** [comment]
Niels, some good jazz on, and I created the dead-simplest repo model to see if it would fail.

It did. Same deletion of the pathed repo. Same style of report forom composer. 

Here is the repo: https://bitbucket.org/narrationsd/compoveri
It's public, as: git@bitbucket.org:narrationsd/compoveri.git

Steps to repeat

1. git clone into folder compoveri
1.  cd compoveri top level
1. composer install. Now you'll have the minimal /vendor folder, with a-repo in it.
2. edit a-repo/composer.json, up the version to 0.1.2
1. composer update
1. Boom. You'll get the message about not being able to reach the pathed repo.
1. the entire a-repo folder will be gone.
1. recover via git: discard changes, pull from origin
1. ready to go again.

That is, if it explodes for you. I'm beginning to have my doubts, but am fresh out of ideas to try, other than to move vendor/composer/composer out for safekeeping, and move my cloned repo in, where it could be instrumented. To quote the famous comedian 'Jose Jimenez, (again) 'Oh, I hope not...'

Cheers, and zzz now.

**Comment by narration-sd** [comment]
Jazz/blues/acoustic samples of last hour, if you guys might enjoy any of it:
- https://www.youtube.com/watch?v=6AVvRqKAzYA
- https://www.youtube.com/watch?v=jAZ7f2uy3ZY
- https://www.youtube.com/watch?v=fwkvC2FT7Xs

**Comment by curry684** [comment]
Well uhm I'm sorry but it's not exploding here:
```
C:\Users\Niels Keurentjes\Projects\compoveri>php ../composer/bin/composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 1 install, 0 updates, 0 removals
  - Installing narra/a-repo (0.1.1): Junctioning from plugin-dev/a-repo
Generating autoload files

C:\Users\Niels Keurentjes\Projects\compoveri>php ../composer/bin/composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating narra/a-repo (0.1.1 => 0.1.2): Junctioning from plugin-dev/a-repo

Writing lock file
Generating autoload files
```
Haven't a clue what would cause the different behavior here. 

Out of curiosity: your project folder is not on any kind of network share? Junctions are *only* possible in NTFS folders. It could most definitely cause funnies when used on SMB/CIFS or NFS shares....

**Comment by narration-sd** [comment]
Well, thanks for checking -- I thought it probably would not.

Here's what I'm about at the moment -- one question to help:
- required your full FileSystem object in the test script, and found isSymLinkedDirectory() also to be reporting appropriately, so that's not the problem.
- figured out your  pathing, and got my clone of origin/composer.git to function
- have xdebug working on it with the local php7.1
- just now looking for the approriate place to breakpoint Filesystem on its runup to doing the bad thing under 'composer update'.

The question: you have three lines dealing with XdebugHandler early in bin/composer. These lead to calling for a restart, which breaks the PhpStorm xdebug connection. I finally just commented them out. Am I going to have trouble from this?

As well, might as well ask where you propose breakpointing FileSystem to be early enough to catch the bad action on 'composer update'. If sure I can find that, good to hear from the author, surely.

Not obsessed with this, nor having much more time to deal with it, but it seems important, and to feed back to you whatever debug may find.

Thanks,
Clive

**Comment by narration-sd** [comment]
Oh, spotted your question. No, no networks involved. All of this is on the prime C: -- Samsung SSD, hyperreliable. But see just above for more info...

**Comment by narration-sd** [comment]
@curry684 @Seldaek 

Ok, guys, I've with some  confidence found and fixed this. 

Let's see if you like it any better than I do.

tl;dr:  changing the `clearstatcache(true, $junction);` on line 667 of your current Util/Filesystem.php to `clearstatcache(false);` will give an apparently dependable fix. 

I can run the repaired Composer and always get local-pathed repo updates, without any removals of junctioned source.

What's this all about?
- it appears to be a timing problem, or possibly order of execution in some wilder imagination
- I could see it happen once only, in PhpStorm xdebug on my clone of your current source, for which I have a picture.
- it happens 100% of the time, running that same clone but without xdebug, on the tiny setup I gave you as repo above.
- it never happens, in my pathck.php checker, even when require-using your Filesystem object.

- _what_ happens is that you get a false mode reply from lstat, causing isJunction to fail. This in turn causes removeDirectory to fall through, and do an unsafe delete through the actual junction. Which has been my problem.

- in fact, instrumenting, whether in my pathck or your object, shows that phplstat _never_ returns a proper value for a junction. 
- When it fails due to the clearstatcache/timing problem, it often returns the signature of an ordinary direcctory I believe, 16895/0x41ff -- but not always. 
- with the fix in place in your object, or at any time in my checker, lstat returns 0 for the junction, consistently. This  fits with the or to your bits coming up zero also, thus isJunction() then functions. It's also entirely the wrong value tested to give that result.

Conclusion: php lstat and its companion clearstatcache are in trouble, at least on the latest  php7.1. The tiny patch appears it would be harmless, as it would work if lstat or clearstatcache were operating properly. 

Performance would be the remaining question, but unless you know better, iit seems calls to isJunction will be of limited proportion even in a massive repository update, and so fully clearing the stat cache there wouldn't be visible.

I didn't think producing a PR worth it for this one-liner; and anyway, you guys have to see if you approve.

Why does it happen for me, this pseudo-random return from lstat()?  I can only guess it's because of a quite fast machine with four cores, possibly in some way tied to using the nts (non-thread-safe) version of php x64, as seems always recommended.. One could in a darker imagination consider about possibly mistaken influences on use of processor caches due to the rushed-out recent security updates. Neither of these explains why when lstat()'s not giving other file/folder's cached permission modes, it's giving straight zeros for junctions, though. 

I suspect we would have to ask our php-dev cousins about that, so, good morning

**Comment by Seldaek** [comment]
If that clearstatcache fixes it I'd say why not, I doubt it would have a significant performance impact.

**Comment by curry684** [comment]
I'm also fine with the change, the explanation is plausible and the stat cache is known to cause issues (I didn't put it there in the first place because it was useless). Performance impact will be negligible, the FS drivers also cache that information more reliable and I've oft wondered why PHP caches it at all - it's a remnant of the days of mechanical hard drives and causes more issues than it fixes these days.

**Comment by narration-sd** [comment]
Ok, then, your PR coming up...

As I'm just back from my constitutional walk, and won't sleep yet.

**Comment by narration-sd** [comment]
There you go.  I would just add that I checked the rather unusual path being presented for isJunction() to test, as it contains both forward and backward slashes. This appears to be accepted actually, presumably due to yeoman Posix compliance measures. 

It made no difference to this #7025 lstat() cache problem when I in-line substituted a fully backslashed Windows style path. Only the clearstatcache(false) patch just submitted did the trick.

**Comment by curry684** [comment]
Congrats on becoming a Composer contributor   😉

**Comment by narration-sd** [comment]
Greetings again. I presume you will have noticed that a test failed in your CI system.

I'd looked at that test last night, and this morning realize that it is likely not a problem in the test jig, which looks a lot like my tester here. 

Of course  there is 'no way' this test should fail, but it does.

I''m going to modify my PR, first putting back in the old code to prove the tester in some sense, and then try the version I first use, which _adds_ the general cache clear after the specific one.

Even if this 'fixes' things, for me the situation as far as the stat cache is worrisome, and I suspect your thought of entirely blocking use of Windows junctions until or unless NTFS symlinks may be indicated -- with a command line-alerted backoff to copying if the filesystem in use doesn't support that.

In its Unix origination, [l]stat used to be a simple kernel system call returning a data structure on the filesystem (inode) -- and it would be considered fundamental.  How it's gotten to have a misbehaving cache at least on the Windows-php combination seems a pretty careless situation, imagine you agree..

**Comment by curry684** [comment]
Don't worry too much about tests failing. Travis is Linux so won't be testing the changed code at all, and Appveyor is pretty much a steaming pile of poo that always fails randomly in my experience.

**Comment by narration-sd** [comment]
Ha - I kind of had my suspicions :)

I think it was Travis that was failing, and at first guessed the Linux aspect, but woke up thinking the platform check should have avoided the problem?? Maybe that didn't function somehow.

Anyway, having both statcacheclear methods in there does pass, should be safe, and might be belt-and-suspenders against the shadow of this php implementation of lstat(), no? If honestly it could go the other way also, but if your and my tests  both pass it?

I'll rerun my test multiply later today, but am pretty confident, and then I'd be most comfortable with the PR to be as it is now, with both methods. You?

**Comment by narration-sd** [comment]
Ok, yes, this ordering is solid here, ten out of ten tries on my test setup, with the two forms of cache clearing in place as the PR has now.

I walk away now...
